documentation/projects/proposals/search_relevancy_sandbox/20230530-implementation_plan_staging_elasticsearch_reindex_dags.md

```diff
@@ -0,0 +1,192 @@
+[delete_index]:
+  https://github.com/WordPress/openverse/blob/7427bbd4a8178d05a27e6fef07d70905ec7ef16b/catalog/dags/data_refresh/data_refresh_task_factory.py#L222-L239
+
+<!--------------------------------------------------------------------------->
+
+### `create_proportional_by_provider_<media>_index` DAG
+
+This DAG is intented to be used most likely with the index resulting from the
+previous DAG or from the data refresh process, that is, an index with the
+database fully indexed, as the `source_index` for the ES
+[Reindex](es_reindex_api) API.
+
+[es_reindex_api]:
+  https://www.elastic.co/guide/en/elasticsearch/reference/7.12/docs-reindex.html
+
+#### Parameters
+
+1. `media_type`: The media type for which the index is being created. Presently
+   this would only be `image` or `audio`.
+2. `source_index`: (Optional) The existing index on Elasticsearch to use as the
+   basis for the new index. If not provided, the index aliased to
+   `<media_type>-full` will be used.
+3. `percentage_of_production`: The proportion of items to take from each
+   provider from the total amount existing in production.
+
+#### Outlined Steps
+
+1. Get the list of media count by sources from the production Openverse API
+   [`https://api.openverse.engineering/v1/<media>/stats/`](https://api.openverse.engineering/v1/<media>/stats/)
+2. Calculate the total media adding up all the counts by provider
+3. Calculate the name of the new index with the following format:
+
+```python
+ f"{media_type}-{percentage_of_production}-percent-of-providers-{current_datetime}"
+```
+
+4. Make a dictionary mapping all the providers with the required amount of items
+   for the new index based on the provided `percentage_of_production` param.
+5. Iterate over the items of the resulting dictionary to index the subset of
+   each provider.
+
+```
+POST _reindex?wait_for_completion=false
+```
+
+```json
+{
+  "max_docs": num_items,
+  "source": {
+    "index": "image-full",
+    "query": {
+      "term": {
+        "source.keyword": "stocksnap"
+      }
+    }
+  },
+  "dest": {
+    "index": "image-50-percent-of-providers-20230608183000"
+  }
+}
+```
+
+6. Make the alias `<media>-subset-by-provider` point to the new index.
+7. Optionally. Query the stats of the resulting index and print the results.
+
+```
+GET /image-reindexed-by-provider/_stats
+```
+
+## Alternatives
+
+### Combining both DAGs into one
+
+One alternative to creating two different indices by separate is to create the
+proportional by provider index using the Ingestion server. This would require
+modifying the REINDEX task of the ingestion server or creating a new one that
+takes only a subset of the providers by the indicated proportion.
+
+However, I discarded this option in favor of the one explained above because
+having both DAGs is much simpler and provides more possibilities for the
+creation of different indexes, which is the end goal of the project.
```

Comment:
:+1: 

Do these DAGs relate to the [index settings iteration DAG](https://docs.openverse.org/projects/proposals/search_relevancy_sandbox/20230518-implementation_plan_staging_index_rapid_iteration.html) at all? Do you think it's worth adding the ability for us to create a proportional index that also has modified settings? From my perspective these proportional indexes are primarily useful for testing different index settings, because it would allow us to "test drive" the modified settings without needing to wait for the entire reindex to happen. I almost wonder if the proportional index creation should be added to the settings iteration DAG as an option :thinking:

Reply:
That's a good idea! I think one approach might be using the already-created proportional index as a baseline for creating a new index with new settings. That _should_ be possible with the DAGs as-defined already I believe!

Reply:
>  I think one approach might be using the already-created proportional index as a baseline for creating a new index with new settings

But why have separate steps for them? It requires _two_ reindexes rather than one. If the DAGs were combined then a single reindex can run, pushing documents into the index with new settings. To get only one or the other behaviour, just leave out one of the arguments.

I don't see why we would need to split this into two distinct steps, is all, unless you're talking about creating a new index without any documents in it. But then the DAG would need to check if we need to create a new index and also be sure that we're not somehow accidentally reindexing into a pre-existing index _accidentally_ instead of intentionally.

Combining them reduces the number of DAGs and the complexity of all of this. The DAG would just be:

1. Pull the source index settings
2. Get modified settings (noop if there are no settings changes passed to the DAG config)
3. Create the new index with the new settings
4. Reindex, either fully or proportionally depending on the DAG config

This appears to be much simpler than two separate DAGs, especially if it can skip a full reindex that we don't always need _especially when testing new settings_. Like I said in the project proposal, I still don't see a very good use case for these proportional indexes _except for testing new index settings_. The features appear to be fundamentally related such that having separate DAGs for them makes the intention of them extremely unclear.

Reply:
I also thought about combining both DAGs into only one. The result of having two comes from that, at first, it seemed like a lot of complexity due to all the moving parts, but now that we know more about ES's APIs and the DAGs are better defined, I would say there is a possibility to combine them.

I still decided to write the proposal as two extra DAGs in case the separation was preferred for flexibility/simplicity and because the `create_new_es_index_{environment}` DAG is still not implemented, so, in theory, this shouldn't depend on that to be implemented. We can also merge them after both are implemented and tested separately.

Reply:
Your perspective makes sense Sara. I'm fine with separate for now with the idea that we can combine them down the line to simplify once things are implemented and tested!